src/metabase/api/collection.clj

```diff
@@ -36,46 +36,90 @@
    [metabase.util.schema :as su]
    [schema.core :as s]
    [toucan.db :as db]
-   [toucan.hydrate :refer [hydrate]]))
+   [toucan.hydrate :refer [hydrate]]
+   [toucan.models :as models]))
 
 ;;; when alias defined for namespaced keywords is run through kondo macro, ns should be regarded as used
 (comment collection.root/keep-me)
 
 (declare root-collection)
 
+(defn- coll-query
+  "Generate a query to return collections viewable by the current user.
+  The user-id is required as well as a boolean flag indicating whether other
+  users' collections (including subcollections) should be ignored (e.g. if the
+  specified user is an admin). Additional where clauses can be provided and
+  will be and'ed in on the where.
+
+  Note that and-clauses should be aliased as c (for collection).
+
+  Also, MySQL 5.7 does not support CTEs so this was all inlined. A much more
+  readable solution existed with CTEs, but we needed to inline for backwards
+  compatibility. TODO: Use CTEs when we drop MySQL 5.7."
+  [owner-id exclude-other-user-collections? & and-clauses]
+  {:select    [:c.*]
+   :from      [[:collection :c]]
+   :left-join [[{:select [[:cwo.id :id]
+                          [true :visible]]
+                 :from   [[{:select    [[:c.id :id]
+                                        [[:coalesce
+                                          :c.personal_owner_id
+                                          :pr.personal_owner_id] :owner]]
+                            :from      [[:collection :c]]
+                            :left-join [[{:select [:c.personal_owner_id
+                                                   [[:concat "/" :c.id "/%"] :pc_prefix]]
+                                          :from   [[:collection :c]]
+                                          :where  [:not= :c.personal_owner_id nil]} :pr]
+                                        [:like :c.location :pr.pc_prefix]]} :cwo]]
+                 :where  [:or
+                          [:= :cwo.owner owner-id]
+                          [:= :cwo.owner nil]]} :a]
+               [:= :a.id :c.id]]
+   :where     (cond-> (into [:and] (or and-clauses [true]))
+                      exclude-other-user-collections? (conj [:= :a.visible true]))
+   :order-by  [[:%lower.name :asc]]})
+
```

Comment:
Note that one design consideration in this was to branch on exclude-other-user-collections? (since it needs the join logic) or just have one query that may not use the join. I opted for the single query as it makes the code more cohesive.